Inflates Account and Container from Json Metadata

xiahome opened the pull request:
This PR introduces versioning and understanding of account/container
metadata, and inflates Account and Container objects from Json metadata.

This PR is split from PR #624 to ease the reviewing work. PR #624 is
still kept open as a reference to the final goal, and will be rebased on
each smaller PRs and merged the last.

Test done: ./gradlew clean && ./gradlew build

Test coverage:
Class Class, %  Method, % Line, %
Account 100% (1/ 1) 92.9% (13/ 14)  92.2% (59/ 64)
Container 100% (1/ 1) 91.7% (11/ 12)  85.2% (52/ 61)

cgtz commented on ambry-api/src/main/java/com.github.ambry/config/HelixPropertyStoreConfig.java:
```diff
@@ -0,0 +1,82 @@
+   */
+  @Config(HELIX_PROPERTY_STORE_PREFIX + "zk.client.connect.string")
+  @Default("")
+  public final String zkClientConnectString;
```
Could we use the connect strings from ClusterMapConfig here?

xiahome replied:
So far there is no problem to simply use the connect strings from the `ClusterMapConfig`, just in theory this is not necessarily the same as the `ClusterMapConfig`. For example, for the purpose of balancing the zk load, one zk is used for cluster management, and one zk is used for account service storage and notifications.

cgtz commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,17 +1,6 @@
-# this file except in compliance with the License. You may obtain a copy of the
-# License at  http://www.apache.org/licenses/LICENSE-2.0
-#
-# Unless required by applicable law or agreed to in writing, software distributed
```
Should we add this copyright message back?

xiahome replied:
Yes, will add back. Not sure how this was automatically removed by IngelliJ.

xiahome commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,263 @@
+   * Gets the id of the account.
+   * @return The id of the account.
+   */
+  public short getId() {
```
Agree, and this is the reason to change it back for consistency.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,263 @@
+ */
+public class Account {
+  // static variables
+  public static final String ACCOUNT_METADATA_VERSION_KEY = "version";
```
do these need to be public ?

xiahome replied:
I think it is not necessary to expose this. Will fix.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,263 @@
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
was this auto generated ?

xiahome replied:
Yes it is auto generated.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,263 @@
+   * @param container The container to add.
+   * @throws JSONException
+   */
+  public void addContainerAndMetadata(Container container) throws JSONException {
```
why public ?

nsivabalan replied:
addOrUpdateContainer(). Metadata is implicit as we don't have over loaded public methods which adds containers.

xiahome replied:
This will usually be called in the account updator's thread. If there is a possibility of multiple updators, there is a need to apply a lock here.

xiahome replied:
This method could be called by an entity to update an `Account`, which is not necessarily in this package.

cgtz replied:
Even if there is a single updator in a different thread than the reader, we would need synchronization, right? I think we might have discussed this offline, but are new containers added to an Account object after it has been provided to a caller via AccountService? If not, than I think this should be a private method and an Account should be immutable once constructed (including its containers). If so, we should look at this more.

xiahome replied:
It should be the second case, i.e., updates to an account can be made through this method. I have some thoughts on synchronization, and will come to you offline for discussion.

xiahome replied:
Will rename.

xiahome replied:
Renamed to `updateContainer`

xiahome replied:
Have updated the patch to introduce a `ReadWriteLock`, which is applied with the write lock when the account is being updated. For read path, the read lock is applied for the mutable fields such as metadata, status, etc.

xiahome replied:
Have made the two classes immutable, so there is no need for synchronization.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,263 @@
+ *    "status": "active"
+ *  }
+ */
+public class Account {
```
did you think about having an interface for Account and Container and defining an implementation for our use-case ? Or what went into deciding not having an interface. Lets say someone wants to come up with a different structure of Containers, but keep the account as is? How do we support such use-cases

xiahome replied:
In the initial design, both `Account` and `Container` are abstracted as interface. The reason to concretize it is they are data structure rather than function components. Versioning and the adoption of json metadata provide a flexible way to extend to a different structure for the two classes. In your example, the different structure of `Container` can be made by changing  the metadata of the `Container` only, and inserting the metadata into the metadata of `Account`. Constructing and parsing `Container` metadata is handled by `Container` itself.

nsivabalan replied:
got you

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,263 @@
+ * Version 1 of account metadata in JSON is in the format below.
+ *  {
+ *    "accountId": 101,
+ *    "accountName": "MyAccount",
```
don't we need a description for an account ?

xiahome replied:
Just compare this to account concept in other systems, normally it is not required to have a description at the account level, but rather for each sub categories. There is no hurt to add description, just not seeing too much necessity. If a use case exists, we can consider adding it.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,263 @@
+  public static final String ACCOUNT_NAME_KEY = "accountName";
+  public static final String ACCOUNT_STATUS_KEY = "status";
+  public static final String CONTAINERS_KEY = "containers";
+  public static final String ACCOUNT_STATUS_ACTIVE = "active";
```
can we define enums for life cycle states?

xiahome replied:
For both `Account` and `Container` class, their status are made as String and some values are predefined as constant. It should be ok for now coz the status is not in a closed definition so far, and there is no complex logic to process the states. Will be moving to a more rigorous definition if needed.

xiahome replied:
Actually have taken your suggestion to use `enum` for the status.

nsivabalan replied:
since this is changed to ENUM, we can represent the status via a short rather than string in the JSON format.

xiahome replied:
Prefer to use string instead of short, so it is more human-readable if we browse the data on zk.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,263 @@
+  public static final String CONTAINERS_KEY = "containers";
+  public static final String ACCOUNT_STATUS_ACTIVE = "active";
+  public static final String ACCOUNT_STATUS_INACTIVE = "inactive";
+  public static final short ACCOUNT_METADATA_VERSION_1 = 1;
```
shouldn't we start with 0?

xiahome replied:
Just to be consistent with other Ambry versioned component like `BlobId`.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,263 @@
+    if (container == null) {
+      throw new IllegalArgumentException("Container cannot be null.");
+    }
+    short accountMetadataVersion = (short) metadata.getInt(ACCOUNT_METADATA_VERSION_KEY);
```
why do we need this check here. An account can be created only by either of two constructors
Account(JSONObject metadata)
public Account(short id, String name, String status)
in both cases, only valid versions will results in creation of account object. Can you clarify on this

xiahome replied:
It is not for checking validity of version, but rather a dispatcher for processing different versions. Now we have only version 1, but in the future we may have version 2, 3, 4...

nsivabalan replied:
got you

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Container.java:
```diff
@@ -0,0 +1,226 @@
+    if (status != null ? !status.equals(container.status) : container.status != null) {
+      return false;
+    }
+    if (description != null ? !description.equals(container.description) : container.description != null) {
```
does description need to be part of equals?

xiahome replied:
I think so, it will be odd if two containers considered to be equal only if their descriptions are not the same.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Container.java:
```diff
@@ -0,0 +1,226 @@
+        this.description = metadata.optString(CONTAINER_DESCRIPTION_KEY);
+        this.isPrivate = metadata.getBoolean(CONTAINER_IS_PRIVATE_KEY);
+        this.metadata = metadata;
+        this.parentAccount = parentAccount;
```
what is the control flow wrt adding a new container to an Acccount. I was expecting parentAccount.addContainerAndMetadata(this) here. Or is the caller expected to call 
```
Account account1= new Account(.....)
Container container1 = new Container(..., account1)
account1.addContainerAndMetadata(container1);
```

xiahome replied:
You are correct.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Container.java:
```diff
@@ -0,0 +1,226 @@
+  public static final String CONTAINER_STATUS_KEY = "status";
+  public static final String CONTAINER_DESCRIPTION_KEY = "description";
+  public static final String CONTAINER_IS_PRIVATE_KEY = "isPrivate";
+  public static final String CONTAINER_STATUS_ACTIVE = "active";
```
again, can we have an enum for the states. Also, do add description as to what each state means

xiahome replied:
Have taken your suggestion.

xiahome commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,263 @@
+   * Gets all the containers of this account in a list.
+   * @return All the containers of this account.
+   */
+  public List<Container> getAllContainers() {
```
It seems in other places our code returns `List` as well, such as get `DataNodeId` and `PartitionId` in `ClusterMap`. I am inclined to returning a list.

cgtz replied:
Okay, that makes sense.

xiahome commented:
@nsivabalan @cgtz Have updated the pr according to your comments online and offline. PTAL.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/AccountBuilder.java:
```diff
@@ -0,0 +1,216 @@
+   * @param container The new {@link Container} to set.
+   * @return This builder.
+   */
+  public AccountBuilder setContainer(Container container) {
```
addOrUpdateContainer() might be a good name

nsivabalan replied:
thinking about this more, I feel that should we have two different methods, one to add a new Container and another to update an existing container. Bcoz, if someone adds a new container w/ an already existing id, we may not catch it or throw an exception. Having two diff methods might make it tight. withing addContainer(Container), if containerId already exists, we should throw an exception.

xiahome replied:
I was trying some code to implement this, but then realized that this might cause some complexity to the caller. If, for example `addContainer`, a new container cannot be added if the same container id already exists, then the same logic should be applied for `updateContainer`, i.e., an exception needs to be thrown if the container does not exist. Then, the caller side may need to have many try-catch blocks and eventually leads to the same goal, add or update the container. So my feeling is a simple `addOrUpdateContainer()` might suffice, and it will not create too much confusion if it is well documented.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/AccountBuilder.java:
```diff
@@ -0,0 +1,216 @@
+   * @param containers A list of {@link Container}s to set.
+   * @return This builder.
+   */
+  public AccountBuilder setContainers(List<Container> containers) {
```
same as above

cgtz replied:
Is this method required? It seems like it's the same as calling setContainer multiple times. I could see it potentially being useful if it would clear the existing list of containers and replace them with the provided list

xiahome replied:
I have removed this method for now.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/AccountBuilder.java:
```diff
@@ -0,0 +1,216 @@
+   * @return This builder.
+   */
+  public AccountBuilder setContainer(Container container) {
+    if (container != null) {
```
shouldn't we do the checkAccountIdInContainers() validation here and throw exception if need be

xiahome replied:
I thought this before, however the situation could be complex. It would be possible to set a container with good parent account id, and then call `setId` to set the parent countainer another id. So finally I feel place the checker in `build()` method would be the best place.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/AccountBuilder.java:
```diff
@@ -0,0 +1,216 @@
+    if (container != null) {
+      short id = container.getId();
+      Container containerToRemove = idToContainerMetadataMap.get(id);
+      if (container.equals(containerToRemove)) {
```
containerToRemove could be null here.

xiahome replied:
`container.equals(null)` should be ok.

nsivabalan replied:
Do you have tests for null case ?

xiahome replied:
Will add

xiahome replied:
It is actually covered by `testRemoveNonExistContainer`.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/AccountBuilder.java:
```diff
@@ -0,0 +1,216 @@
+    checkRequiredFields(version);
+    switch (version) {
+      case ACCOUNT_METADATA_VERSION_1:
+        if (id == null) {
```
this is repetition of code in checkRequiredFields()

xiahome replied:
Removed.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/AccountBuilder.java:
```diff
@@ -0,0 +1,216 @@
+  private void checkRequiredFields(short version) {
+    switch (version) {
+      case CONTAINER_METADATA_VERSION_1:
+        if (id == null) {
```
I see that the constructor do take in all 3 args. Shouldn't we do this validation during construction itself and fail for any such invalid values.

xiahome replied:
My thought is that anything can be changed before `build()` is called. So `build()` should be the final place to make all the checks (check required fields, consistency between `accountId` and container's `parentAccountId`).

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,240 @@
+   * Adds a {@link Container} to this account and updates internal maps accordingly.
+   * @param container The container to update this account.
+   */
+  private void updateContainerMap(Container container) {
```
I don't think we need a separate method for this. we can make it inline.

cgtz replied:
Agreed, this is only called in one place now

xiahome replied:
removed.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,240 @@
+        }
+        metadata.put(CONTAINERS_KEY, containerArray);
+        break;
+
```
why line break?

xiahome replied:
Just for readability.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/AccountBuilder.java:
```diff
@@ -0,0 +1,216 @@
+        jsonMetadata.put(ACCOUNT_METADATA_VERSION_KEY, version);
+        jsonMetadata.put(ACCOUNT_ID_KEY, id);
+        jsonMetadata.put(ACCOUNT_NAME_KEY, name);
+        jsonMetadata.put(ACCOUNT_STATUS_KEY, status);
```
are we not using the enum's ordinal rather than the enum's string representation

cgtz replied:
Since JSON isn't super size efficient to begin with, I think it's better to use the string representation. This means that we can more easily modify the AccountStatus enum in the future without requiring that the position of the existing enum values stay the same

xiahome replied:
Adding more, using a zk inspector it would be more human-readable.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,240 @@
+   */
+  @Override
+  public String toString() {
+    return "Account[" + getId() + "]";
```
why not even printing account name

xiahome replied:
The reason is to make it to be consistent with `toString()` for container. In the container's version, the container instance does not have account name information, only account id.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Container.java:
```diff
@@ -0,0 +1,213 @@
+   */
+  @Override
+  public String toString() {
+    return "Container[" + getParentAccountId() + ":" + getId() + "]";
```
why not printing container name

xiahome replied:
The same reason as explained above.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/ContainerBuilder.java:
```diff
@@ -0,0 +1,194 @@
+    this.version = CONTAINER_METADATA_VERSION_1;
+  }
+
+  /**
```
this java docs seems out of place. Did you forget to remove it

xiahome replied:
oops, removed.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/AccountBuilder.java:
```diff
@@ -0,0 +1,216 @@
+    for (Container container : origin.getAllContainers()) {
+      idToContainerMetadataMap.put(container.getId(), container);
+    }
+    this.version = ACCOUNT_METADATA_VERSION_1;
```
something to think about. Incase we introduce a new version for Account sometime in future, how will this code look like. Some fields in V2 may not be available in V1. Do we have dummy values for it ?

xiahome replied:
We will always build the highest version objects. So if we have v2, we need to supply dummy values for those deserialized from v1 json, which indicates N/A.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/ContainerBuilder.java:
```diff
@@ -0,0 +1,194 @@
+    if (origin == null) {
+      throw new IllegalArgumentException("origin cannot be null.");
+    }
+    this.id = origin.getId();
```
same question as above. How do we handle multiple versions of Container passed in as an argument

xiahome replied:
You have brought a good point. The logic in `AccountBuilder` and `ContainerBuilder` has been changed to always construct into an object in the highest version, and always serialize into json in the highest version.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/ContainerBuilder.java:
```diff
@@ -0,0 +1,194 @@
+          throw new IllegalStateException("Cannot build account, name is not set");
+        }
+        if (status == null) {
+          throw new IllegalStateException("Cannot build account, status is not set");
```
I dont think we need this level of detailed logging. You can do the following
if(id == null || name == null || stats == null || ........................................................)
      throw new IllegalArgumentException("Either of name "+name+" or id "+ id +" ......." is null);

if you fix this, do the same in case of AccountId as well

xiahome replied:
Changed as suggested.

cgtz commented on ambry-api/src/main/java/com.github.ambry/account/AccountBuilder.java:
```diff
@@ -0,0 +1,216 @@
+   * @param id The id to set.
+   * @return This builder.
+   */
+  public AccountBuilder setId(short id) {
```
One thing to think about is that changing the account id would make the parent account id in any child containers invalid. We could either recreate all of the containers currently in `idToContainerMetadataMap`, or we could get rid of this method, since it is a required field from the constructor anyways.

xiahome replied:
As discussed offline, would prefer to do the check in the `build()` method, which is the final stage to build. Anything could change before this stage. Once a checker is there, we do not need to worry about any intermediate stage, the final `build()` call will either succeed or fail.

xiahome commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -175,54 +207,69 @@ public Container getContainerByContainerId(short containerId) {
    * @throws JSONException
    */
-  public void addContainerAndMetadata(Container container) throws JSONException {
+  public void updateContainer(Container container) throws JSONException {
```
Not valid after the new commit.

xiahome commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -175,54 +207,69 @@ public Container getContainerByContainerId(short containerId) {
+          readWriteLock.writeLock().unlock();
+        }
         break;
 
```
For beautification.

xiahome commented:
@nsivabalan @cgtz Have pushed a new commit. Mainly addressed three issues:
1. `Account/Container` always serialized into the highest metadata version;
2. Moved validation to constructors;
3. Added more validation for `container` inconsistency.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,312 @@
+          for (int index = 0; index < containerArray.length(); index++) {
+            JSONObject containerMetadata = containerArray.getJSONObject(index);
+            Container container = new Container(containerMetadata);
+            containers.add(container);
```
can we do the parent check here before adding it to the map. Also we can do the duplicate check too here itself. 
```
if (container.getParentAccountId() != id) {
        throw new IllegalStateException(
            "Container does not belong to this account because parentAccountId=" + container.getParentAccountId()
                + " is not the same as accountId=" + id);
  }
if(containerIdToContainerMap.containsKey(container.getId()) || containerNameToContainerMap.containsKey(container.getName()){
               throw exception....
  }
```

xiahome replied:
I got your point to do the checks inside of the for loop. Will come to you and discuss if this would be a better place. I made some attempt and there is pros and cons in both way.

xiahome replied:
Have introduced `checkParentAccountIdInContainers(Container container)` and `checkDuplicateContainerNameOrId(Container container)` and called within the `for` loop.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,312 @@
+   * @param status The status of the account. Cannot be null.
+   * @param containers A collection of {@link Container}s to be part of this account.
+   */
+  Account(Short id, String name, AccountStatus status, Collection<Container> containers) {
```
shouldn't we make this private. I mean, move AccountBuilder to this file and make this constructor private so that no one else can construct an Account instance apart from builder

xiahome replied:
I was trying to do the same, however, changed to the current shape for two reasons:
1. `Account(JSONObject metadata)` will be called by `AccountService`, which is outside of the class. We could do something extra like `AccountBuilder.buildFromJson(JSONObject metadata)`, so we can completely hide the two constructors;
2. `account` package so far is relatively isolated and safe, but we can discuss if 1 is the right move.

xiahome replied:
Have introduced `public static Account fromJson(JSONObject json)` to deserialize a `JSONObject` to `Account` object, which is within the `Account` class. Also made the `Account(JSONObject obj)` constructor private. However, as discussed offline, we kept `Account` and `AccountBuilder` two separate classes.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,312 @@
+  static final String ACCOUNT_STATUS_KEY = "status";
+  static final String CONTAINERS_KEY = "containers";
+  static final short ACCOUNT_METADATA_VERSION_1 = 1;
+  static final short HIGHEST_ACCOUNT_METADATA_VERSION = ACCOUNT_METADATA_VERSION_1;
```
can you name this CURRENT_VERSION. Also for other variables, why do we need "metadata" prefixes. Just "version" should suffice since anyone would call it as "Account.Version" or "Account.Current_Version" sort of

xiahome replied:
Have renamed to `CURRENT_JSON_VERSION`, because this is for the serialized json version, not the account class itself.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Container.java:
```diff
@@ -0,0 +1,243 @@
+  static final String CONTAINER_DESCRIPTION_KEY = "description";
+  static final String CONTAINER_IS_PRIVATE_KEY = "isPrivate";
+  static final String CONTAINER_PARENT_ACCOUNT_ID_KEY = "parentAccountId";
+  static final short CONTAINER_METADATA_VERSION_1 = 1;
```
same comments as Account

xiahome replied:
Renamed to `CONTAINER_JSON_VERSION_1`

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Container.java:
```diff
@@ -0,0 +1,243 @@
+   * @return The String representation of this container.
+   */
+  @Override
+  public String toString() {
```
can we have another method to print all details for a container. We might need it for trace logging. Same for account.

xiahome replied:
Seems `toJson()` is the one we need.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,312 @@
+            updateContainerMap(container);
+          }
+          checkAccountIdInContainers();
+          checkDuplicateContainerNameOrId(containers);
```
shouldn't we call into other constructor once we have all the fields.

cgtz replied:
I think it would be cleaner too to do these checks in the other constructor, but I think the call to another constructor has to be the first statement in this constructor's body, so idk how to make that work with the json parsing.

xiahome replied:
Guess we could not, because constructor needs to be called at the first place.

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/Container.java:
```diff
@@ -0,0 +1,243 @@
+    short metadataVersion = (short) metadata.getInt(CONTAINER_METADATA_VERSION_KEY);
+    switch (metadataVersion) {
+      case CONTAINER_METADATA_VERSION_1:
+        this.id = (short) metadata.getInt(CONTAINER_ID_KEY);
```
shouldn't we call into other constructor once we have all fields.

cgtz replied:
Same comment here. Idk if this is possible in java

nsivabalan commented on ambry-api/src/main/java/com.github.ambry/account/ContainerBuilder.java:
```diff
@@ -0,0 +1,138 @@
+   * @param parentAccountId The parent {@link Account} id to set.
+   * @return This builder.
+   */
+  public ContainerBuilder setParentAccountId(Short parentAccountId) {
```
I see this has not been used. Guess you might wanna have a test to cover this

xiahome replied:
Added a unit test.

cgtz commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,312 @@
+      for (Container container : containers) {
+        updateContainerMap(container);
+      }
+      checkAccountIdInContainers();
```
I think these two check methods could be combined into a single validateContainers(Collection<Container>) method instead of iterating twice.

cgtz replied:
Nevermind, I just saw that checkDuplicateContainerNameOrId() does not actually iterate over the containers

cgtz commented on ambry-api/src/main/java/com.github.ambry/account/Account.java:
```diff
@@ -0,0 +1,312 @@
+   * @param containers A collection of containers to check duplicate name or id.
+   * @throws IllegalStateException if there are containers that have different ids but the same names, or vise versa.
+   */
+  private void checkDuplicateContainerNameOrId(Collection<Container> containers) {
```
Could this method just take in the expectedNumberOfContainers? This means that we won't have to construct a new list when building an account from JSON metadata.

xiahome replied:
Have made the method to call against each individual `Container`.

cgtz commented on ambry-api/src/main/java/com.github.ambry/account/Container.java:
```diff
@@ -0,0 +1,243 @@
+  private final String name;
+  private final ContainerStatus status;
+  private final String description;
+  private final Boolean isPrivate;
```
minor: can we use primitives for short and boolean instead of the object wrappers? Same comment in the Builders and Account. Then we wouldn't have to do the null checks for these (since null isn't an available value for primitives).

xiahome replied:
Would prefer to have a single place to check all the required fields, and `name` fields will anyway be checked in the constructors.

xiahome commented:
@nsivabalan @cgtz Have pushed a new commit for the changes.

nsivabalan commented on ambry-api/src/test/java/com.github.ambry/account/AccountContainerTest.java:
```diff
@@ -0,0 +1,574 @@
+ */
+public class AccountContainerTest {
+  // Reference Account fields
+  private final short refAccountId = 1234;
```
can you generate random short for every test.

xiahome replied:
Done.

nsivabalan commented on ambry-api/src/test/java/com.github.ambry/account/AccountContainerTest.java:
```diff
@@ -0,0 +1,574 @@
+public class AccountContainerTest {
+  // Reference Account fields
+  private final short refAccountId = 1234;
+  private final String refAccountName = "testAccount1";
```
same here. Random strings and random values for every field possible.

xiahome replied:
Done.

nsivabalan commented on ambry-api/src/test/java/com.github.ambry/account/AccountContainerTest.java:
```diff
@@ -0,0 +1,574 @@
+      Arrays.asList(new String[]{"Public container", "Private container"});
+  private final List<ContainerStatus> refContainerStatuses =
+      Arrays.asList(new Container.ContainerStatus[]{ContainerStatus.ACTIVE, ContainerStatus.INACTIVE});
+  private final List<Boolean> refContainerPrivacies = Arrays.asList(new Boolean[]{false, true});
```
refContainerPrivacyValues

xiahome replied:
Renamed

nsivabalan commented on ambry-api/src/test/java/com.github.ambry/account/AccountContainerTest.java:
```diff
@@ -0,0 +1,574 @@
+  public void testConstructAccountAndContainerFromArguments() throws JSONException {
+    List<Container> containers = new ArrayList<>();
+    for (int i = 0; i < 2; i++) {
+      containers.add(new Container(refContainerIds.get(i), refContainerNames.get(i), refContainerStatuses.get(i),
```
shouldn't we use Builder here. coz, no one outside the package will have access to constructor.

xiahome replied:
Have changed to use the builder.

nsivabalan commented on ambry-api/src/test/java/com.github.ambry/account/AccountContainerTest.java:
```diff
@@ -0,0 +1,574 @@
+   * @param metadata An invalid metadata in Json.
+   * @param exceptionClass The class of expected exception.
+   */
+  private void assertCreatAccountWithBadJsonMetadata(JSONObject metadata, Class<?> exceptionClass) {
```
assertAccountCreationFailure

xiahome replied:
Renamed to 'createAccountWithBadJsonMetadata'. Same naming has been applied to other similar methods.

nsivabalan commented on ambry-api/src/test/java/com.github.ambry/account/AccountContainerTest.java:
```diff
@@ -0,0 +1,574 @@
+   * @param status The status for the {@link Account} to build.
+   * @param exceptionClass The class of expected exception.
+   */
+  private void assertBuildAccountWithMissingFields(Short id, String name, AccountStatus status,
```
assertBuildAccountFailure

xiahome replied:
Renamed

nsivabalan commented on ambry-api/src/test/java/com.github.ambry/account/AccountContainerTest.java:
```diff
@@ -0,0 +1,574 @@
+    }
+
+    // remove two existing containers.
+    for (Container container : origin.getAllContainers()) {
```
is there a test, wherein you remove just one container and test if the other exists. Ideally, if you have N containers, remove random M (<n) containers and validate. Also remove all N and validate.

nsivabalan commented on ambry-api/src/test/java/com.github.ambry/account/AccountContainerTest.java:
```diff
@@ -0,0 +1,574 @@
+      fail("should have thrown");
+    } catch (Exception e) {
+      System.out.println(e.getMessage());
+      assertEquals("Wrong exception", exceptionClass, e.getClass());
```
why do we need to print the msg here? You can add the msg in the assertion failure msg.

xiahome replied:
Removed all the `System.out.println`

xiahome commented:
@nsivabalan Made a new commit to address your comments.

nsivabalan commented on ambry-api/src/test/java/com.github.ambry/account/AccountContainerTest.java:
```diff
@@ -489,12 +489,11 @@ private void assertContainer(Container container, int index) throws JSONExceptio
    * @param exceptionClass The class of expected exception.
    */
-  private void assertCreateAccountWithBadContainers(List<Container> containers, Class<?> exceptionClass) {
+  private void createAccountWithBadContainers(List<Container> containers, Class<?> exceptionClass) {
```
I would expect the method name to depict that we expect failure.

xiahome replied:
Followed your suggestion and renamed to `***AndFail`

nsivabalan commented on ambry-api/src/test/java/com.github.ambry/account/AccountContainerTest.java:
```diff
@@ -340,7 +338,6 @@ public void testRemoveContainerAndUpdateAccount() throws JSONException {
-      System.out.println(e.getMessage());
     }
 
     // remove two existing containers.
```
in line 343, you are removing all containers. Also, as I suggested, can you remove M no of containers randomly. Also after each removal, check for non existence of the container from the account ref. In the end, you can remove all and do all of your validations.

xiahome replied:
I have added a new unit test for removing containers.

nsivabalan commented on ambry-api/src/test/java/com.github.ambry/account/AccountContainerTest.java:
```diff
@@ -373,15 +370,15 @@ public void testUpdateContainerInAccount() throws JSONException {
       Container container = account.getContainerById(refContainerIds.get(i));
       accountBuilder.removeContainer(container);
       ContainerBuilder containerBuilder = new ContainerBuilder(container);
       short updatedContainerId = (short) (container.getId() + 100);
```
how do we know that +100 may not collide with any other existing container Ids ?

xiahome replied:
Good point! Have fixed.

xiahome replied:
Changed to `short updatedContainerId = (short) (-1 * (container.getId()));`, since all the randomly-generated container ids are non-negative;

xiahome commented:
@nsivabalan Have made changes according to your comments.
